hyperspy/drawing/spectrum.py

```diff
@@ -331,7 +331,16 @@ def update(self, force_replot=False):
             ydata = self.data_function(axes_manager=self.axes_manager).real
         else:
             ydata = self.data_function(axes_manager=self.axes_manager).imag
-        self.line.set_ydata(ydata)
+
+        new_axis = self.axes_manager.signal_axes[0].axis
+        if len(new_axis) != len(self.axis) or \
+           (len(new_axis) == len(self.axis) and
+                np.any(np.not_equal(new_axis, self.axis))):
+            self.axis = new_axis
+            self.ax.set_xlim(self.axis[0], self.axis[-1])
+        self.line._xorig = self.axis
+        self.line._yorig = ydata
+        self.line.recache(True)
```

Comment:
That's essentially how they do it inside, and it's the only way to update both `x` and `y` without a `recache` running between the two runs (which returns an error if the lengths are different) that I've found...